lib/CleanPlugin.js

```diff
@@ -0,0 +1,224 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Sergey Melyukov @smelukov
+*/
+
+"use strict";
+
+const asyncLib = require("neo-async");
+const path = require("path");
+const { validate } = require("schema-utils");
+const { SyncHook } = require("tapable");
+const Compilation = require("../lib/Compilation");
+const schema = require("../schemas/plugins/CleanPlugin.json");
+const { join } = require("./util/fs");
+
+/** @typedef {import("../declarations/plugins/CleanPlugin").CleanPluginArgument} CleanPluginArgument */
+/** @typedef {import("./Compiler")} Compiler */
+
+/** @typedef {(function(string):boolean)|RegExp} IgnoreItem */
+/** @typedef {function(IgnoreItem): void} AddToIgnoreCallback */
+
+/**
+ * @typedef {Object} CleanPluginCompilationHooks
+ * @property {SyncHook<[AddToIgnoreCallback]>} ignore
+ */
+
+/** @type {WeakMap<Compilation, CleanPluginCompilationHooks>} */
+const compilationHooksMap = new WeakMap();
+
+/**
+ * @param {IgnoreItem} ignore regexp or function
+ * @param {string} asset asset path
+ * @returns {boolean} true if an asset should be ignored
+ */
+const checkToIgnore = (ignore, asset) => {
+	if (ignore instanceof RegExp) {
+		return !!ignore.exec(asset);
+	}
+
+	return ignore(asset);
+};
+
+class CleanPlugin {
+	/**
+	 * @param {Compilation} compilation the compilation
+	 * @returns {CleanPluginCompilationHooks} the attached hooks
+	 */
+	static getCompilationHooks(compilation) {
+		if (!(compilation instanceof Compilation)) {
+			throw new TypeError(
+				"The 'compilation' argument must be an instance of Compilation"
+			);
+		}
+		let hooks = compilationHooksMap.get(compilation);
+		if (hooks === undefined) {
+			hooks = {
+				/** @type {SyncHook<[AddToIgnoreCallback]>} */
+				ignore: new SyncHook(["ignore"])
+			};
+			compilationHooksMap.set(compilation, hooks);
+		}
+		return hooks;
+	}
+
+	/** @param {CleanPluginArgument} [options] options */
+	constructor(options) {
+		this.options = Object.assign({ enabled: true, dry: false }, options || {});
+
+		if (options && typeof options === "object") {
+			validate(schema, options, {
+				name: "Clean Plugin",
+				baseDataPath: "options"
+			});
+		}
```

Comment:
I think we should avoid getting non object options, anyway if we will accept non object options we need validate it too

Reply:
We accept only object options
`output.clean` may accept `boolean | object`